### ISSUE TITLE
Avoid setting __proto__

### ISSUE THREAD

**Issue by NTillmann** [facebook/prepack#574]
When an object needs a custom prototype, Prepack currently generates an assignment like this:
var obj = { ... }
// possibly more property assignments 
obj.__proto__ = ...

Setting __proto__ after object creation is an expense operations on most modern JavaScript engines. I did some measuring, found some performance degradation on v8 on micro benchmarks, and I found an insane performance degradation on JSC. The following program
```
(function() {

function f(x, y, prev) {
    this.x=y;
    this.y=y;
    this.prev = prev;
}

function fEmptyConstructor() {
}
fEmptyConstructor.prototype = f.prototype;

function createViaConstructor(x, y, prev) {
    return new f(x, y, prev);
}

function createViaEmptyConstructor(x, y, prev) {
    var obj = new fEmptyConstructor();
    obj.x = x;
    obj.y = y;
    obj.prev = prev;
    return obj;
}

function createVia__proto__(x, y, prev) {
    var obj = { x: x, y: x, prev: prev };
    obj.__proto__ = f.prototype;
    return obj;
}

function measure(name, f) {
  var start = Date.now();
  var prev;
  for (var i=0; i<1000000; i++) { prev = f(1,2,prev); }
  var end = Date.now();
  print(name + ": " + (end-start));
}

measure("constructor call", createViaConstructor);
measure("create via empty constructor", createViaEmptyConstructor);
measure("create via __proto__", createVia__proto__);

})();
```
prints 
```
onstructor call: 523
create via empty constructor: 1466
create via __proto__: 12735
```
when running with an old version of JSC!

So, as a quick fix that will make things ten times as fast, let's emit the "create with an empty auxiliary constructor wired up to the right prototype" pattern. Long term, we should refactor our "factorify" logic, and emit proper auxiliary constructors that set all the fields right away when there are a lot of instances of a particular pattern.

**Comment by samwgoldman** [comment]
What about using `__proto__` in the object initializer, as described in [the ES2015 spec](https://www.ecma-international.org/ecma-262/6.0/#sec-__proto__-property-names-in-object-initializers). AIUI, this is non-standard before ES2015, but was standardized because everyone seems to support it.

**Comment by cblappert** [comment]
Tested this by adding
```javascript
function createViaLiteral(x, y, prev) {
  var obj = { x: x, y: x, prev: prev, __proto__: f.prototype };
  return obj;
}
```
 results are:
```
constructor call: 630
create via empty constructor: 1775
create via obj literal with __proto__: 19955
create via __proto__: 20924
```
Numbers are a bit higher because I'm running on a slower machine.

**Comment by samwgoldman** [comment]
Sounds like `__proto__` in the initializer is a non-solution :)

**Comment by hermanventer** [comment]
I had a look at the places where this is occurs in the scenario we are currently most interested in and it appears to be something a non issue because only class functions and prototype objects get their __proto__ properties set in the prepacked code. I'm lowering the priority of this for the time being.

**Comment by cblappert** [comment]
Solved by #667
